File: Tutorials/Tuckshop/Client/Tuckshop.WebApp/src/Domain/Models/Orders/Commands/CreateOrder.ts
import { Attributes, List, ModelBase, Rules, Validation } from '@singularsystems/neo-core';

export class CreateOrder extends ModelBase {
    static typeName = "CreateOrder";

    constructor() {
        super();
        this.makeObservable();
    }

    @Rules.StringLength(100)
    public customerName: string = "";

    public orderDetails = new List(NewOrderDetail);

    // Client only properties / methods

    protected static addBusinessRules(rules: Validation.Rules<CreateOrder>) {
        super.addBusinessRules(rules);
    }

    public toString(): string {
        if (this.isNew || !this.customerName) {
            return "New create order";
        } else {
            return this.customerName;
        }
    }
}

export class NewOrderDetail {

    public productId: number = 0;

    @Attributes.Integer()
    public quantity: number = 0;

    // Client only properties / methods
}

File: Tutorials/Tuckshop/Server/Tuckshop.Domain/Tuckshop.Core.Models/Models/Orders/Commands/CreateOrder.cs
namespace Tuckshop.Core.Models.Orders.Commands
{
  using System.Collections.Generic;
  using System.ComponentModel.DataAnnotations;
  using Neo.Model;
  using Neo.Model.Validation;

  /// <summary>
  /// Class representing a new order command.
  /// </summary>
  public class CreateOrder : ModelBase<CreateOrder>
  {

    /// <summary>
    /// Gets or sets the Customer Id. Null when IsCashSale is true.
    /// </summary>
    public int? CustomerId { get; set; }

    /// <summary>
    /// Gets and sets a value indicating whether this is a cash sale (one-ime / walk-in customer).
    /// </summary>
    public bool IsCashSale { get; set; }

    /// <summary>
    /// Gets and sets the Customer Name.
    /// </summary>
    [Required]
    [MaxLength(100)]
    public string CustomerName { get; set; } = string.Empty;

    /// <summary>
    /// Gets and sets the Order Details.
    /// </summary>
    public ICollection<NewOrderDetail> OrderDetails { get; set; } = new List<NewOrderDetail>();

    /// <summary>
    /// Represents a new order detail, which includes the product ID  and the Quantityfor the order.
    /// </summary>
    public class NewOrderDetail : ModelBase<NewOrderDetail>
    {
      /// <summary>
      /// Gets or sets the Order Detail Id.
      /// </summary>
      public int ProductId { get; set; }

      /// <summary>
      /// Gets or sets the Quantity.
      /// </summary>
      public int Quantity { get; set; }
    }

    /// <inheritdoc/>
    protected override void AddBusinessRules(ValidationRules<CreateOrder> rules)
    {
      rules.FailWhen(
      ord => ord.OrderDetails == null || ord.OrderDetails.Count == 0,
      "Order details are required");

      rules.FailWhen(
        ord => !ord.IsCashSale && ord.CustomerId == null,
        "A customer must be selected unless this is a cash sale.");
    }
  }
}
